### PR TITLE
ceph: option to enable/disable host networking in CSI pods

### LINKED ISSUE #7203: Disable hostnetworking for CSI drivers
currently, cephcsi daemonset pods are using the hostnetworking by default. with https://github.com/ceph/ceph-csi/pull/1450 we don't need hostnetworking anymore for the RBD driver and the CephFS driver is already working fine without hostnetworking (tested during multus integration).

We can provide an option to disable the hostnetworking for now. For the existing cluster after the upgrade cephcsi will still use the hostnetworking (to avoid issues if something is present). 

If anyone wants to try out without hostnetworking we can have an option (`CSI_DISABLE_HOSTNETWORK`) to disable hostnetworking. Maybe after a few cycles, we can disable hostnetworking by default for CSI pods.

Note:-   For this feature, we need a new cephcsi release. This PR is not part of (v3.2.1)

cc @idryomov @dillaman @travisn @leseb @nixpanic

### PR DESCRIPTION
**Pull request by Rakshith-R**
**Description of your changes:**

As host networking is no longer necessary for CSI driver, an option CSI_ENABLE_HOST_NETWORK is added to 
`rook-ceph-operator-config` configmap to enable/disable host network.
Helm chart docs, values.yaml and deployment.yaml template is also edited to support the change.

- New Clusters will have host networking set to be disabled in the CSI driver.
- Upgraded Clusters will default to host networking enabled in the CSI driver. 

Note:- Needs a new cephcsi release

Signed-off-by: Rakshith R <rar@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #7203 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: cluster/charts/rook-ceph/templates/psp.yaml [hunk 1]
```diff
@@ -58,8 +58,6 @@ spec:
   hostIPC: true
   # Ceph OSDs need to share the same PID namespace
   hostPID: true
-  # hostNetwork can be set to 'false' if host networking isn't used
```
Comment (Madhu-1) [inline cluster/charts/rook-ceph/templates/psp.yaml:61]: this not related to cephcsi. we should not remove it.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
we need to have an option to enable/disable hostnetworking which is missing here.

details are captured  at https://github.com/rook/rook/issues/7203#issuecomment-789795095

File: cluster/charts/rook-ceph/templates/psp.yaml (hunk 1 above)
Comment (Rakshith-R) [reply cluster/charts/rook-ceph/templates/psp.yaml:61]: okay

**Comment by Rakshith-R** [comment]
Hey @Madhu-1 , updated the pr. Please take a look, Thanks.

File: Documentation/helm-operator.md [hunk 2]
```diff
@@ -110,6 +110,7 @@ The following tables lists the configurable parameters of the rook-operator char
 | `csi.cephFSFSGroupPolicy`          | Policy for modifying a volume's ownership or permissions when the CephFS PVC is being mounted                                   | ReadWriteOnceWithFSType                                  |
 | `csi.logLevel`                     | Set logging level for csi containers. Supported values from 0 to 5. 0 for general useful logs, 5 for trace level verbosity.    | `0`                                                       |
 | `csi.enableGrpcMetrics`            | Enable Ceph CSI GRPC Metrics.                                                                                                  | `false`                                                   |
+| `csi.enableCSIHostNetwork`         | Enable Host Networking for Ceph CSI nodeplugins.                                                                               | `true`                                                   |
```
Comment (Madhu-1) [inline Documentation/helm-operator.md:113]: default is false?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 3]
```diff
@@ -105,6 +105,9 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   ROOK_CSI_ENABLE_GRPC_METRICS: "false"
 
+  # set to true to enable host networking for CSI CEPHFS and RBD nodeplugins.
```
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:108]: change CEPHFS to CephFS. take care of it in other places also

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -54,6 +54,7 @@ type Param struct {
 	RBDPluginUpdateStrategy      string
 	PluginPriorityClassName      string
 	ProvisionerPriorityClassName string
+	EnableCSIHostNetwork         string
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:57]: can this be bool?

File: pkg/operator/ceph/csi/spec.go [hunk 5]
```diff
@@ -229,6 +231,7 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 	RBDDriverName = tp.DriverNamePrefix + "rbd.csi.ceph.com"
 
 	tp.EnableCSIGRPCMetrics = fmt.Sprintf("%t", EnableCSIGRPCMetrics)
+	tp.EnableCSIHostNetwork = fmt.Sprintf("%t", EnableCSIHostNetwork)
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:234]: if we declare the `EnableCSIHostNetwork` we don't need this one.

**Comment by Madhu-1** [comment]
added DNM till we get the cephcsi release which is having required fix for the rbd noudev CLI option in rbd map

File: pkg/operator/ceph/csi/csi.go [hunk 6]
```diff
@@ -85,6 +85,14 @@ func SetParams(clientset kubernetes.Interface) error {
 		return errors.Wrap(err, "unable to parse value for 'ROOK_CSI_ENABLE_GRPC_METRICS'")
 	}
 
+	csiEnableCSIHostNetwork, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_ENABLE_HOST_NETWORK", "true")
```
Comment (travisn) [inline pkg/operator/ceph/csi/csi.go:88]: Default is false, right?
```suggestion
	csiEnableCSIHostNetwork, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_ENABLE_HOST_NETWORK", "false")
```

File: cluster/examples/kubernetes/ceph/operator.yaml [hunk 7]
```diff
@@ -29,6 +29,9 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   ROOK_CSI_ENABLE_GRPC_METRICS: "false"
 
+  # set to true to enable host networking for CSI CEPHFS and RBD nodeplugins.
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator.yaml:32]: Do we know of a scenario where it might need to be enabled? If not, let's at least mention in a comment when they might need to. What about this?
```suggestion
  # Set to true to enable host networking for CSI CEPHFS and RBD nodeplugins. This may be necessary in some network configurations where the SDN does not provide access to an external cluster.
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by Rakshith-R** [comment]
- `CSI_ENABLE_HOST_NETWORK` needs to default to `true` to enable host networking for upgraded clusters.
- Example `rook-ceph-operator-config` configmap and helm `values.yaml` contain `CSI_ENABLE_HOST_NETWORK: false ` so that new clusters will have host networking disabled 

I've made other changes pointed out above. 
PTAL @Madhu-1 @travisn  Thanks

**Comment by travisn** [comment]
> * `CSI_ENABLE_HOST_NETWORK` needs to default to `true` to enable host networking for upgraded clusters.
> * Example `rook-ceph-operator-config` configmap and helm `values.yaml` contain `CSI_ENABLE_HOST_NETWORK: false ` so that new clusters will have host networking disabled
> 
> I've made other changes pointed out above.
> PTAL @Madhu-1 @travisn Thanks

I'm thinking the risk of turning off host networking is low and we can just always default to disabling it by default, no need to default to true for upgraded clusters. Please add a note to PendingReleaseNotes.md to mention this change, and we can make sure to highlight it in case they see issues with the external clusters they can enable it. If we learn otherwise before the v1.6 release, we can change the default, but I'd like to assume it's not necessary until proven otherwise.

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 8]
```diff
@@ -105,6 +105,10 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   ROOK_CSI_ENABLE_GRPC_METRICS: "false"
 
+  # Set to true to enable host networking for CSI CephFS and RBD nodeplugins.This may be necessary
+  # in some network configurations where the SDN does not provide access to an external cluster.
```
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]: @travisn provisioner pod is able to reach the monitor IPs in external mode and provisioner pod is already running on pod network. do we need to rephrase this one?

Comment (travisn) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]: Sure, what is your suggestion on the wording?

File: pkg/operator/ceph/csi/csi.go [hunk 9]
```diff
@@ -84,6 +84,14 @@ func SetParams(clientset kubernetes.Interface) error {
 		return errors.Wrap(err, "unable to parse value for 'ROOK_CSI_ENABLE_GRPC_METRICS'")
 	}
 
+	csiEnableCSIHostNetwork, err := k8sutil.GetOperatorSetting(clientset, controllerutil.OperatorSettingConfigMapName, "CSI_ENABLE_HOST_NETWORK", "false")
+	if err != nil {
+		return errors.Wrap(err, "unable to determine if CSI Host Network is enabled")
```
Comment (leseb) [inline pkg/operator/ceph/csi/csi.go:89]:
```suggestion
		return errors.Wrap(err, "failed to determine if CSI Host Network is enabled")
```

Rook is using `failed to` across its code base to produce error message.

File: pkg/operator/ceph/csi/csi.go [hunk 10]
```diff
@@ -84,6 +84,14 @@ func SetParams(clientset kubernetes.Interface) error {
...
+	}
+	if CSIParam.EnableCSIHostNetwork, err = strconv.ParseBool(csiEnableCSIHostNetwork); err != nil {
+		return errors.Wrap(err, "unable to parse value for 'CSI_ENABLE_HOST_NETWORK'")
```
Comment (leseb) [inline pkg/operator/ceph/csi/csi.go:92]:
```suggestion
		return errors.Wrapf(err, "failed to parse value for 'CSI_ENABLE_HOST_NETWORK'")
```

**Comment by travisn** [comment]
@Rakshith-R Please rebase on the latest so we can see if that fixes the CI issues. It would be good to get this merged in the next day or two so we can test for the v1.6 release.

**Review by travisn — COMMENTED** [review]
Looks like the CI still needs investigation

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 8 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]: as am thinking ,maybe if someone sees read/write performance drop with this change can revert back to hostnetworking as earlier

File: pkg/operator/ceph/csi/csi.go (hunk 10 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/csi.go]: @leseb we don't need Wrapf here, right?

Comment (leseb) [reply pkg/operator/ceph/csi/csi.go]: No we don't indeed, I guess I was thinking we could pass the var from  `CSI_ENABLE_HOST_NETWORK` but nevermind :)

File: cluster/charts/rook-ceph/values.yaml [hunk 11]
```diff
@@ -91,7 +94,7 @@ csi:
   # Default value is RollingUpdate.
   #cephFSPluginUpdateStrategy: OnDelete
   # Allow starting unsupported ceph-csi image
-  allowUnsupportedVersion: false
+  allowUnsupportedVersion: true
```
Comment (leseb) [inline cluster/charts/rook-ceph/values.yaml:97]: Why is this needed? Is it leftover?

File: cluster/examples/kubernetes/ceph/operator.yaml [hunk 12]
```diff
@@ -60,7 +64,7 @@ data:
   CSI_CEPHFS_FSGROUPPOLICY: "ReadWriteOnceWithFSType"
 
   # (Optional) Allow starting unsupported ceph-csi image
-  ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+  ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "true"
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/operator.yaml:67]: ditto?

File: pkg/operator/ceph/csi/spec.go [hunk 13]
```diff
@@ -105,7 +106,7 @@ var (
 // manually challenging.
 var (
 	// image names
-	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:v3.2.1"
+	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:canary"
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:109]: Can't we just wait for the upcoming 3.3 release which should go with Rook 1.6?

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 13 above)
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/spec.go:109]: > Can't we just wait for the upcoming 3.3 release which should go with Rook 1.6?

@leseb  Those changes were just to check everything worked with host network disabled. I will change them back after cephcsi 3.3 release.

**Comment by Madhu-1** [comment]
@travisn looks like we forgot to take this in for 1.6 release, can we target for next release as csi 3.3.0 is having the support

**Comment by travisn** [comment]
We will plan on v1.6.1 later next week, sounds good to include it there

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 8 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]: @Rakshith-R still this needs to be addressed?

**Review by Madhu-1 — APPROVED** [review]
changes LGTM. just had one suggestion related to comment https://github.com/rook/rook/pull/7356/files#r596628272

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 8 above)
Comment (Rakshith-R) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]:
``` 
  # Set to true to enable host networking for CSI CephFS and RBD nodeplugins.This may be necessary
  # in some network configurations where the SDN does not provide access to an external cluster or
  # there is significant drop in read/write performance.
```
@Madhu-1  This sounds good ?

Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:109]: Looks good

File: PendingReleaseNotes.md [hunk 14]
```diff
@@ -37,3 +37,4 @@ v1.6...
 * Monitors failover can be [disabled](Documentation/ceph-mon-health.md#failing-over-a-monitor), this is useful for monitor going under planned maintenance where automatic failover is not desired
 * RGW: Rook has started using deployment's replicatset functionality instead of deploying multiple deployments for each rgw
 * Support running Volume Replication Controller on the RBD provisioner pod. Depends on Cephcsi v3.3.0
+* Disable Host Network in CSI CephFS and RBD nodeplugin pods by default. `CSI_ENABLE_HOST_NETWORK` in `rook-ceph-operator-config` configmap and `enableCSIHostNetwork` in helm values.yaml can be set to true to enable Host Network
```
Comment (leseb) [inline PendingReleaseNotes.md:40]: We don't need that anymore since the 1.6 is out, I just reset the file here https://github.com/rook/rook/pull/7692. This will be included in the 1.6.1 note but not in the `PendingReleaseNotes.md` file.

Comment (leseb) [inline PendingReleaseNotes.md:40]: We don't need that anymore since the 1.6 is out, I just reset the file here https://github.com/rook/rook/pull/7692. This will be included in the 1.6.1 note but not in the `PendingReleaseNotes.md` file.

File: cluster/charts/rook-ceph/values.yaml [hunk 15]
```diff
@@ -56,6 +56,10 @@ csi:
   enableRbdDriver: true
   enableCephfsDriver: true
   enableGrpcMetrics: false
+  # Set to true to enable host networking for CSI CephFS and RBD nodeplugins.This may be necessary
```
Comment (leseb) [inline cluster/charts/rook-ceph/values.yaml:59]: `s/nodeplugins.This/nodeplugins. This/`

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 16]
```diff
@@ -105,6 +105,11 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   ROOK_CSI_ENABLE_GRPC_METRICS: "false"
 
+  # Set to true to enable host networking for CSI CephFS and RBD nodeplugins.This may be necessary
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:108]: ditto

**Review by leseb — CHANGES_REQUESTED** [review]
A few nits

**Review by leseb — APPROVED** [review]
